indexers/keyvalue/RedisDBIndexer/tests/test_redisdbindexer.py

```diff
@@ -1,58 +1,205 @@
+        elif method == 'update':
+            idx.update(keys, documents_bytes)
+        else:
+            print(f'method {method} is not implemented')
+        idx.touch()
+        idx.save()
+
+
+def apply_actions(save_abspath, index_abspath, actions):
+    for action in actions:
+        idx = BaseIndexer.load(save_abspath)
+        apply_action(idx, action)
+
+    return save_abspath, index_abspath
+
+
+def validate_positive_results(keys, documents, searcher):
+    for key, query_doc in zip(keys, documents):
+        result_doc = Parse(searcher.query(key)[0]['values'].decode('utf8'), Document())
+        assert result_doc.id == str(query_doc[0]) * 16
+        assert result_doc.buffer == query_doc[1].encode('utf8')
+        assert round(result_doc.weight, 5) == query_doc[2]
+        assert result_doc.length == query_doc[3]
+
+
+def validate_negative_results(keys, searcher):
+    for key in keys:
+        result_docs = searcher.query(key)
+        assert result_docs == []
+
+
+def validate_results(save_abspath, results, negative_results):
+    with BaseIndexer.load(save_abspath) as searcher:
+        if results:
+            keys, ids = zip(*[[k, v] for k, v in results.items()])
+            _, documents, _ = get_documents(ids)
+            validate_positive_results(keys, documents, searcher)
+        if negative_results:
+            validate_negative_results(negative_results, searcher)
+
+
+def run_crud_test_for_scenario(indexer, actions, results, no_results, tmpdir):
+    import redis
+    r = redis.Redis(host='0.0.0.0', port=63079, db=0, socket_timeout=10)
+    r.flushdb()
+    indexer.workspace = tmpdir
+    with indexer:
+        indexer.touch()
+        indexer.save()
+    save_abspath = indexer.save_abspath
+    index_abspath = indexer.index_abspath
+    assert Path(index_abspath).exists()
+    assert Path(save_abspath).exists()
+    apply_actions(save_abspath, index_abspath, actions)
+    validate_results(save_abspath, results, no_results)
+
+
+def run_crud_test(actions, results, no_results, tmpdir):
+    # test construction from code
+    indexer = RedisDBIndexer(level='doc', db=0)
+    run_crud_test_for_scenario(indexer, actions, results, no_results, tmpdir)
+
+    # test construction from yaml
+    from jina.executors import BaseExecutor
+    indexer = BaseExecutor.load_config(str(cur_dir / 'yaml/test-redis.yml'))
+    run_crud_test_for_scenario(indexer, actions, results, no_results, tmpdir)
+
+
+def run_crud_test_exception_aware(actions, results, no_results, exception, mocker, tmpdir):
+    # action is defined as (method, key, document_id)
+    if exception is not None:
+        with pytest.raises(exception):
+            run_crud_test(actions, results, no_results, tmpdir)
+    else:
+        global validate_positive_results, validate_negative_results
+        validate_positive_results = mocker.Mock(wraps=validate_positive_results)
+        validate_negative_results = mocker.Mock(wraps=validate_negative_results)
```

Comment:
When you moved the other functions as suggested, this must not be an overwrite via `global` anymore, but can become
```
_validate_positive_results = mocker.Mock(wraps=validate_positive_results
...
validate_results(..., _validate_positive_results, _validate_negative_results)
```
which is way safer.

Reply:
Nice, I changed it to function parameters.